### ISSUE TITLE
feature proposal: lxc network: vlan support

### ISSUE THREAD

**Issue by FransM** [lxc/lxd#3002]
I was wondering whether it wouldn't be nice to have vlan support build in into lxc network.
I want my containers to be fully isolated from each other. Currently what I do is create a vlan for each container on the host using vconfig.

After that I can create a network (bridge) with this new vlan interface and attach my container network interface.

I am creating the containers remotely and now need an additional ssh to the host to create the vlan interface. It would be nice if lxc could also do the vlan creation for me (the vconfig call).

(note, I need vlan, not macvlan)

**Comment by stgraber** [comment]
Yeah, I've been thinking about that in the past.

My plan is to add a "vlan" field to the "macvlan" network type so you could do:
 - lxc config device add CONTAINER vlan-1000 nic nictype=macvlan vlan=1000 name=eth0 parent=bond0

Which would then look for a VLAN device for ID 1000 on the host bond0 device. If it finds one, then create a macvlan over that and pass it to the container. If it doesn't, then create a new VLAN device on the host and then do the macvlan trick on top of that.

**Comment by stgraber** [comment]
Note that the way you find a particular existing VLAN device on the host for a given VID and parent device is by parsing /proc/net/vlan/config

**Comment by stgraber** [comment]
We need to use macvlan as otherwise it wouldn't be possible to connect multiple containers to the same VLAN.

**Comment by FransM** [comment]
Actually the idea I had is a bit different from this (but maybe can be realized with this).
I have a number of containers running on the same host. On this host I've manually set up vlan's, one for each container. So I have eth0.1, eth0.2, eth0.3 etc and each container has its own vlan.

Traffic from the host goes to a switch that redirects vlan1 to port 1, vlan2 to port 2 vlan3 to port 3 etc (where the switch removes the tag).

So basically I want to use vlans to give each container its own physical network outlet

**Comment by stgraber** [comment]
Yeah, that'd still work fine, the code will re-use your existing vlan interfaces if they exist or create them if they don't.

Just do:

    lxc config device add c1 eth0 nic nictype=macvlan name=eth0 parent=eth0 vlan=1

And c1 will be bound to eth0.1.
